core/transaction/utils.go

```diff
@@ -0,0 +1,400 @@
+	if minNumConfirmation > len(sharders) {
+		return nil, errors.New("verify_optimistic", "wrong number of min_confirmations")
+	}
+	shuffled := util.Shuffle(sharders)[:minNumConfirmation]
+
+	//prepare urls for confirmation request
+	urls := make([]string, 0, len(shuffled))
+	mappedSharders := make(map[string]string)
+	for _, sharder := range shuffled {
+		url := fmt.Sprintf("%v/%v%v", sharder, TXN_VERIFY_URL, txnHash)
+		urls = append(urls, url)
+		mappedSharders[url] = sharder
+	}
+
+	//initialize resty
+	header := map[string]string{
+		"Content-Type":                "application/json; charset=utf-8",
+		"Access-Control-Allow-Origin": "*",
+	}
+
+	transport := createTransport(resty.DefaultDialTimeout)
+
+	options := []resty.Option{
+		resty.WithTimeout(resty.DefaultRequestTimeout),
+		resty.WithRetry(resty.DefaultRetry),
+		resty.WithHeader(header),
+		resty.WithTransport(transport),
+	}
+
+	var url string
+	var chain []*RoundBlockHeader
+	var txn *Transaction
+	r := resty.New(options...).Then(func(req *http.Request, resp *http.Response, respBody []byte, cf context.CancelFunc, err error) error {
+		if err != nil { //network issue
+			return err
+		}
+
+		if resp.StatusCode != 200 {
+			return errors.Throw(ErrInvalidRequest, strconv.Itoa(resp.StatusCode)+": "+resp.Status)
+		}
+
+		//parse response
+		var objmap map[string]json.RawMessage
+		err = json.Unmarshal(respBody, &objmap)
+		if err != nil {
+			return err
+		}
+		txnRawJSON, ok := objmap["txn"]
+		// txn data is found, success
+		if !ok {
+			return errors.New("handle_response", "bad transaction response")
+		}
+		merklePathRawJSON, ok := objmap["merkle_tree_path"]
+		if !ok {
+			return errors.New("handle_response", "bad merkle_tree_path response")
+		}
+
+		txn = &Transaction{}
+		err = json.Unmarshal(txnRawJSON, txn)
+
+		b := &RoundBlockHeader{}
+		err = json.Unmarshal(respBody, b)
+		if err != nil {
+			return err
+		}
+		err = validateBlockHash(b)
+		if err != nil {
+			return err
+		}
+
+		err = verifyMerklePath(err, merklePathRawJSON, txn.Hash, b.MerkleTreeRoot)
+		if err != nil {
+			return err
+		}
+
+		url = req.URL.String()
+		chain = append(chain, b)
+		return nil
+	})
+
+	retries := 0
+	ticker := time.NewTicker(time.Second)
+L:
+	//loop query confirmation
+	for retries < 30 {
+		select {
+		case <-ticker.C:
+			retries++
+			r.DoGet(context.TODO(), urls...)
+			//need single valid confirmation
+			errs := r.First()
+			if len(errs) == 0 {
+				break L
+			}
+		}
+	}
+
+	if len(chain) == 0 {
+		return nil, errors.New("verify", "can't get confirmation after 10 retries")
+	}
+
+	//remove current sharder from the list to avoid building chain with it
+	toDelete := mappedSharders[url]
+	for i, s := range sharders {
+		if s == toDelete {
+			sharders = append(sharders[:i], sharders[i+1:]...)
+			break
+		}
+	}
+
+	//build blockchain starting from confirmation block
+	curRound := chain[0].Round
```

Comment:
Let's move the confirmation code to a separate one to avoid making this function too long to read